homeassistant/components/shelly/__init__.py

```diff
@@ -24,25 +23,34 @@
 _LOGGER = logging.getLogger(__name__)
 
 
-async def async_setup(hass: HomeAssistant, config: dict):
-    """Set up the Shelly component."""
-    hass.data[DOMAIN] = {DATA_CONFIG_ENTRY: {}}
-    hass.data[DOMAIN][COAP_CONTEXT] = await aiocoap.Context.create_client_context()
+async def get_coap_context(hass):
```

Comment:
Use singleton decorator

```python
from homeassistant.helpers import singleton

@singleton.singleton(COAP_CONTEXT)
async def get_coap_context(hass):
    context = aioshelly.COAP()
    await context.initialize()

    @callback
    def shutdown_listener(ev):
        context.close()

    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, shutdown_listener)

    return context
```

That way it's initialized once, race conditions are handled and you're no longer attaching a shutdown listener each time. 

Do not use `COAP_CONTEXT` anywhere else but by calling the decorator. That's an implementation detail. In other places just do

```python
context = get_coap_context(hass)
```